torch/quantization/fx/prepare.py

```diff
@@ -160,7 +166,13 @@ def insert_observer(
     if model_device:
         observer.to(model_device)
     # add observer module as attribute
-    prefix = node.name + '_activation_post_process_'
+    if isinstance(observer, _InputEqualizationObserver) or isinstance(observer, _WeightEqualizationObserver):
+        prefix = node.name + '_equalization_process_'
+    else:
+        if 'equalization' in node.name:
+            prefix = node.name.replace('equalization', 'activation_post')
```

Comment:
what's the context on this?  Could you give an example of when this code path is taken?

It would be good to avoid replacing parts of strings, it's really convenient to have the entire prefix in the code so that it is easy to grep for.

Reply:
In the original code (without the equalization observers), the graph after prepare for `x -> InputQuantObs` would look something like: `x -> x_activation_post_process_0`. The naming scheme is just taking the name of the previous node (`x`) and adding `_activation_post_process_`.

After the input equalization observer is be inserted before the input quantization observer, the graph after prepare would something like `x -> InputEqObs -> InputQuantObs` and would now be labeled as `x -> x_equalization_process_0 -> x_equalization_process_0_activation_post_process_0`. 

Since this naming is slightly inaccurate as the `InputQuantObs` should be `x`'s activation post process, I tried to get rid of the additional `x_equalization_process_0` in the third term by removing the entire `equalization` part in the name and replacing it with `activation`.

Reply:
makes sense, thanks for explaining.  Might be good to just keep the naming simple (`x_equalization_process_0_activation_post_process_0` sounds ok), it would be easier to debug that way.

that said, are we sure this order is what we want?  I would have expected the quantization observer to come first, and the equalization observer to come second.  For example, if we have `linear1 -> linear2`, we would have `eq_obs1 -> linear1 -> obs1` inserted when processing linear1, and `eq_obs2 -> linear2 -> obs2` inserted when processing linear2.  So, the order between `linear1` and `linear2` is quantization observer first, and equalization observer second.

Reply:
I think the motivation for this was that the equalization observer gets replaced by a mul and the quantization observer gets replaced by a quantize_per_tensor (for fp32 input) and the expected graph after convert is `x -> mul -> quantize_per_tensor -> quantized_lienar -> dequant`, so the prepared graph is `x -> eq_obs -> in_obs -> linear -> out_obs`

For the 2 linear case, I think the order would be `x -> eq_obs -> inp_obs -> linear1 -> out_obs -> eq_obs2 -> linear2 -> out_obs2` (since linear1 out_obs gets inserted first). 
I think this can be changed to have the eq_obs appear after the quantization observer for the first linear (with fp32 input), it shouldn't affect the functionality. @angelayi lmk what you think

Reply:
> that said, are we sure this order is what we want? 

This might be a little confusing, but here are the cases:
- For the input layer, it should look like: `x -> eq_obs -> in_obs -> linear`
- This is this same as when there is a non-quantized operation (assume `add` is a `fp32` op): `add -> eq_obs -> in_obs -> linear`
- For connected layers, it is like you said, `linear1 -> out_obs1 -> eq_obs2 -> linear2 -> out_obs2`. However, in this case we do not need to scale the next input quantization observer so the specific location of the equalization observer is not really necessary. Instead, we can scale `linear1`'s weight value by the next equalization scale (I have a PR coming for this soon).

Reply:
yes, unfortunately in this case (at least in how I think it would be the cleanest to build) the order of observers in the graph is not going to match the order of nodes after convert.  It would be

```
# after prepare
x0 -> qobs0 -> eqobs1 -> linear1

# after convert
x0 -> mul(from eqobs1) -> quantize_per_tensor(from qobs0) -> qlinear1
```

I believe this is better because it allows us to keep the invariant that in prepare, we do not need to modify the graph before the current node in order to insert observers for the current node.

Reply:
> For the 2 linear case, I think the order would be x -> eq_obs -> inp_obs -> linear1 -> out_obs -> eq_obs2 -> linear2 -> out_obs2 (since linear1 out_obs gets inserted first).
> I think this can be changed to have the eq_obs appear after the quantization observer for the first linear (with fp32 input), it shouldn't affect the functionality.

Yes, for the 2 connected layers case, it is currently: `linear1 -> out_obs1 -> eq_obs2 -> linear2 -> out_obs2`. However, in this case we do not need to scale the next input quantization observer so the specific location of the equalization observer (whether it is before or after `out_obs1`) is not really necessary. Instead, we can scale `linear1`'s weight value by the next equalization scale (I have a PR coming for this soon).

I'm not sure how to exactly change the existing code to insert `eq_obs2` before `out_obs1`, but I agree that it shouldn't affect the functionality. The reason why I'm not totally sure how to change the existing code is that I think right now the observers are inserted in the following manner:
1. Initially we have `x -> linear1 -> linear2`
2. We look `linear1` and insert the equalization observer, quantization input observer, and quantization output observer: `x -> eq_obs1 -> in_obs1 -> linear1 -> out_obs1 -> linear2`
3. We look at `linear2` and insert the equalization observer: `x -> eq_obs1 -> in_obs1 -> linear1 -> out_obs1 -> eq_obs2 -> linear2`
4. (still looking at `linear2`), we cannot insert the input equalization observer because it has the same type as the `out_obs1` (we do not want to insert duplicate observers).
If we want to insert the `eq_obs2` before `out_obs1` then we would need to move some of the nodes around (if that makes sense).